### PR TITLE
feat(dist): add notes for `stable` and `beta` in `components_missing_msg()`

### PR DESCRIPTION
**Pull request by rami3l**
Prepares for `rls`' removal in the release channel.

Continuation of #3453, as requested by @onur-ozkan in https://github.com/rust-lang/rust/pull/126856#issuecomment-2197990717.

> The best thing we could do is to add a note saying some components may no longer be available (this note currently doesn't apply for stables), but:
>
> 1. We'll have to wait for v1.28 to ship it, and we're currently in the middle of the release cycle (we have limited bandwidth so we decided not to do backports), and
> 2. We should not hardcode the list of removed components in Rustup
> 
> _https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/excluding.20rls.20from.20the.20release/near/447968780_

~~The error message is not final, since I'd still like to add a link that provides more info about removed components (which is `rls` this time):~~ The message now includes a link to https://rust-lang.github.io/rustup/devel/concepts/components.html#previous-components.

> ... and for this I suggest adding a section in the Rust docs somewhere, and our notes should advise the user to check out the section above to see whether the component in question is missing as intended from what date (and if that is the case, they can choose to unblock the update with a clear consent by removing the component).
>
> But in any way we cannot simply skip RLS in an update if it's previously installed, since that would be a real backwards compatibility problem.
> _https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/excluding.20rls.20from.20the.20release/near/447968780_

## Concerns

- [x] ~~To which document should we add the list of removed components?~~ To the Rustup User Guide, see above for details.
- [x] ~~Should the note be applied to `nightly` as well?~~ No, as https://github.com/rust-lang/rustup-components-history/pull/54 will add a link to https://rust-lang.github.io/rustup/devel/concepts/components.html#previous-components on the components history page.

### CODE REVIEW

File: src/dist/mod.rs
Comment (djc) [inline src/dist/mod.rs]: It's a little hard to say without full context, but I think "these components" would read better?

File: src/dist/mod.rs
Comment (djc) [inline src/dist/mod.rs:106]: The wording here seems a little vague. Can we do something more specific, perhaps by checking for the specific `rls` component?

Comment (rami3l) [reply src/dist/mod.rs:106]: @djc As explained above, I intend to learn from the message right above (regarding nightlies) and to add a link pointing to a list of removed components somewhere online (see the `Concerns` section of the original post). As a result, not mentioning `rls` explicitly here is intentional, since this gives the release team the freedom to remove components in the middle of any Rustup release cycle once this fix lands, as long as the online list can be updated in time.

File: src/dist/mod.rs
Comment (rbtcollins) [inline src/dist/mod.rs]:
```suggestion
If you require the component(s), please install and use the latest successfully built version,
```

File: src/dist/mod.rs
Comment (rami3l) [reply src/dist/mod.rs]: @djc Fixed!

File: src/dist/mod.rs
Comment (rami3l) [reply src/dist/mod.rs:106]: @djc I've decided to use https://rust-lang.github.io/rustup/devel/concepts/components.html#previous-components, and I've updated this message to include that link.

**Comment by rami3l** [comment]
> * [ ]  Should the note be applied to `nightly` as well?

@onur-ozkan My apologies for getting a bit sidetracked previously! I would like to know how https://rust-lang.github.io/rustup-components-history will behave on the removal of `rls`, since we include this link in the message for nightly users.

IMHO, it should also include a link to https://rust-lang.github.io/rustup/devel/concepts/components.html#previous-components in case someone wonders why `rls` is gone in consecutive nightlies starting from a certain date.

**Comment by onur-ozkan** [comment]
> @onur-ozkan My apologies for getting a bit sidetracked previously! I would like to know how https://rust-lang.github.io/rustup-components-history will behave on the removal of rls, since we include this link in the message for nightly users.

I couldn't find any hard-coded logic in https://github.com/rust-lang/rustup-components-history. I think it will just disappear without causing any issues.

> IMHO, it should also include a link to https://rust-lang.github.io/rustup/devel/concepts/components.html#previous-components in case someone wonders why rls is gone in consecutive nightlies starting from a certain date.

That sounds great but I don't know who maintains it.

**Comment by rami3l** [comment]
> > IMHO, it should also include a link to https://rust-lang.github.io/rustup/devel/concepts/components.html#previous-components in case someone wonders why rls is gone in consecutive nightlies starting from a certain date.
> 
> That sounds great but I don't know who maintains it.

cc @Mark-Simulacrum @mexus

**Comment by Mark-Simulacrum** [comment]
I'm happy to merge small PRs to rustup-components-history.

File: src/dist/mod.rs
Comment (djc) [inline src/dist/mod.rs:102]: Should we do `beta-` and `stable-` instead to more precisely match the toolchain name boundary?

**Review by djc — APPROVED** [review]

File: src/dist/mod.rs
Comment (rami3l) [reply src/dist/mod.rs:102]: @djc Hmmm, I don't think what you've suggested will work with `+stable`.

**Pull request merged**
